Update readme to detail the lerobot dataset format (#275)
huggingface/lerobot: README.md

Check out [example 1](./examples/1_load_lerobot_dataset.py) that illustrates how to use our dataset class which automatically download data from the Hugging Face hub.

You can also locally visualize episodes from a dataset on the hub by executing our script from the command line:
```bash
python lerobot/scripts/visualize_dataset.py \
    --repo-id lerobot/pusht \
    --episode-index 0
```

> or from a dataset in a local folder with the root `DATA_DIR` environment variable

Reviewer: Is this a good place to mention something like "this would attempt to load data from `./my_local_data_dir/lerobot/pusht`"? (please double check that is correct)

Author: added

```bash
DATA_DIR='./my_local_data_dir' python lerobot/scripts/visualize_dataset.py \
    --repo-id lerobot/pusht \
    --episode-index 0
```

It will open `rerun.io` and display the camera streams, robot states and actions, like this:

https://github-production-user-asset-6210df.s3.amazonaws.com/4681518/328035972-fd46b787-b532-47e2-bb6f-fd536a55a7ed.mov?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240505%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240505T172924Z&X-Amz-Expires=300&X-Amz-Signature=d680b26c532eeaf80740f08af3320d22ad0b8a4e4da1bcc4f33142c15b509eda&X-Amz-SignedHeaders=host&actor_id=24889239&key_id=0&repo_id=748713144

Our script can also visualize datasets stored on a distant server. See `python lerobot/scripts/visualize_dataset.py --help` for more instructions.

> ### The `LeRobotDataset` format

Reviewer: Perhaps there's a good place somewhere here to mention what policy inputs and outputs should look like?

```
Data keys can vary to some extent between datasets, but one must keep in mind that LeRobot code makes use some structure in the data key naming:

- The action data is always under the "action" key.
- Image observations use either "observation.image" if there is only one image per frame, or "observation.images.{name}" if there are multiple.
- The robot state data is always under the "observation.state" key.
```

Author: yes good point. This is actually a pain point since we have some hidden conversions between environments and policies. Maybe let's figure this out in another PR.

> A dataset in `LeRobotDataset` format is very simple to use. It can be loaded from a repository on the Hugging Face hub or a local folder simply with e.g. `dataset = LeRobotDataset("lerobot/aloha_static_coffee")` and can be indexed into like any Hugging Face and Pytorch dataset. For instance `dataset[0]` will retrieve a sample of the dataset observations and actions in pytorch tensors format ready to be fed to a model.

Reviewer: Maybe a good place to define "sample" (plus other nitty edits)

```suggestion
A dataset in `LeRobotDataset` format is very simple to use. It can be loaded from a repository on the Hugging Face hub or a local folder simply with e.g. `dataset = LeRobotDataset("lerobot/aloha_static_coffee")` and can be indexed into like any Hugging Face and PyTorch dataset. For instance `dataset[0]` will retrieve a single temporal frame from the dataset containing observation(s) and an action as PyTorch tensors ready to be fed to a model.
```

> A specificity of `LeRobotDataset` is that we can retrieve several frames for one sample query. By setting `delta_timestamps` to a list of delta timestamps, e.g. `delta_timestamps = {"observation.image": [-1, -0.5, -0.2, 0]}`  one can retrieve, for each query, 4 images including one at -1 second before the current time step, the two others at -0.5 second and -0.2, and the final one at the current time step (0 second). See example [1_load_lerobot_dataset.py](examples/1_load_lerobot_dataset.py) for more details on `delta_timestamps`.

Reviewer: Suggestion incorporates:

- specificity early on
- don't explain `delta_timestamps` with "delta timestamps"
- "query" -> "index" to stay in line prior terminology
- don't put "the two others" into their own special group 
 
```suggestion
A specificity of `LeRobotDataset` is that, rather than retrieving a single frame by its index, we can retrieve several frames based on their temporal relationship with the indexed frame, by setting `delta_timestamps` to a list of relative times with respect to the indexed frame. For example, with `delta_timestamps = {"observation.image": [-1, -0.5, -0.2, 0]}`  one can retrieve, for a given index, 4 frames: 3 "previous" frames 1 second, 0.5 seconds, and 0.2 seconds before the indexed frame, and the indexed frame itself (corresponding to the 0 entry). See example [1_load_lerobot_dataset.py](examples/1_load_lerobot_dataset.py) for more details on `delta_timestamps`.
```

> Under the hood, the `LeRobotDataset` format makes use of several ways to serialize data which can be useful to understand if you plan to work more closely with this format. We tried to make a flexible yet simple dataset format that would cover most type of features and specificities present in reinforcement learning and robotics, in simulation and in real-world, with a focus on cameras and robot states.

Reviewer: "with a focus on cameras and robot states"

It's super easily extendable to other types of sensory inputs as long as they can be represented by a tensor. Also to environment states (which can be done by cheating in-sim, or by having a perception module).

Here are the important details and internal structure organization of a typical `LeRobotDataset` instantiated with `dataset = LeRobotDataset("lerobot/aloha_static_coffee")`. The exact features will change from dataset to dataset but not the main aspects:

```

> dataset attributes:

Reviewer: I wish there was a way of flagging this snippet so that there's a CI mechanism to draw our attention to it in case we change anything in the datasets module ... or something, anything to make sure it doesn't go stale. I wonder if someone at HF has solved this problem...

Author: Yes indeed, we have that in `transformers` for docstring examples. It's quite q lot of work to setup and maintain so I would advise to wait that the library has really build a large community of users before diving in this.

  ├ hf_dataset: a Hugging Face dataset (backed by Arrow/parquet). Typical features example:

>   │  ├ observation.images.cam_high: VideoFrame
>   │  │   VideoFrame = {'path': path to a mp4 video, 'timestamp': float32 timestamp in the video}
>   │  ├ observation.state: List of float32: position of an arm joints (for instance)

Reviewer 2: nit
```suggestion
  │  ├ observation.images.cam_high (VideoFrame):
  │  │   VideoFrame = {'path': path to a mp4 video, 'timestamp' (float32): timestamp in the video}
  │  ├ observation.state (list of float32): position of an arm joints (for instance)
```

  │  ... (more observations)

>   │  ├ action: List of float32

Reviewer 2: ```suggestion
  │  ├ action (list of float32): goal position of an arm joints (for instance)
```

>   │  ├ episode_index: int64: index of the episode for this sample
>   │  ├ frame_index: int64: index of the frame for this sample in the episode ; starts at 0 for each episode
>   │  ├ timestamp: float32: timestamp in the episode
>   │  ├ next.done: bool: indicates the end of en episode ; True for the last frame in each episode
>   │  └ index: int64: general index in the whole dataset

Reviewer 2: nit
```suggestion
  │  ├ episode_index (int64): index of the episode for this sample
  │  ├ frame_index (int64): index of the frame for this sample in the episode ; starts at 0 for each episode
  │  ├ timestamp (float32): timestamp in the episode
  │  ├ next.done (bool): indicates the end of en episode ; True for the last frame in each episode
  │  └ index (int64): general index in the whole dataset
```

  ├ episode_data_index: contains 2 tensors with the start and end indices of each episode

>   │  ├ from: 1D int64 tensor of first frame index for each episode: shape (num episodes,) starts with 0
>   │  └ to: 1D int64 tensor of last frame index for each episode: shape (num episodes,)

Reviewer 2: nit
```suggestion
  │  ├ from (1D int64 tensor): first frame index for each episode — shape (num episodes,) starts with 0
  │  └ to: (1D int64 tensor): last frame index for each episode — shape (num episodes,)
```

  ├ stats: a dictionary of statistics (max, mean, min, std) for each feature in the dataset, for instance
  │  ├ observation.images.cam_high: {'max': tensor with same number of dimensions (e.g. `(c, 1, 1)` for images, `(c,)` for states), etc.}
  │  ...
  ├ info: a dictionary of metadata on the dataset

>   │  ├ fps: float - frame per second the dataset is recorded/synchronized to
>   │  └ video: bool - indicates if frames are encoded in mp4 video files to save space or stored as png files
>   ├ videos_dir: path to where the mp4 videos or png images are stored/accessed
>   └ camera_keys: List of string: the keys to access camera features in the item returned by the dataset (e.g. `["observation.images.cam_high", ...]`)

Reviewer 2: nit
```suggestion
  │  ├ fps (float): frame per second the dataset is recorded/synchronized to
  │  └ video (bool): indicates if frames are encoded in mp4 video files to save space or stored as png files
  ├ videos_dir (Path): where the mp4 videos or png images are stored/accessed
  └ camera_keys (list of string): the keys to access camera features in the item returned by the dataset (e.g. `["observation.images.cam_high", ...]`)
```

```

A `LeRobotDataset` is serialised using several widespread file formats for each of its parts, namely:
- hf_dataset stored using Hugging Face datasets library serialization to parquet
- videos are stored in mp4 format to save space or png files
- episode_data_index saved using `safetensor` tensor serializtion format

> - stats saved using `safetensor` tensor serializtion format

Reviewer: nit: more than one place btw
```suggestion
- stats saved using `safetensor` tensor serialization format
```

- info are saved using JSON

> Dataset can uploaded/downloaded from the HuggingFace hub seamlessly. To work on a local dataset, you can set the `DATA_DIR` environment variable to you root dataset folder as illustrated in the above section on dataset visualization.

Reviewer: Is this repetition needed?

huggingface-cli login --token ${HUGGINGFACE_TOKEN} --add-to-git-credential
```

> Then move your dataset folder in `data` directory (e.g. `data/aloha_static_pingpong_test`), and push your dataset to the hub with:

Reviewer 2: This looks like a merge/rebase conflict error as we just updated this in #231
```suggestion
Then point to your raw dataset folder (e.g. `data/aloha_static_pingpong_test_raw`), and push your dataset to the hub with:
```
